[DH-483] Adds sector name to Contact.

elcct opened the pull request:
This adds a `sector` from related `Company` object to the `Contact`.

I have added new `COMPUTED_MAPPINGS` map that will contain fields that do not exist directly in source model.

reupen commented on datahub/search/dict_utils.py:
```diff
@@ -32,6 +32,26 @@ def _contact_dict(obj):
+        if field is None:
+            return None
+
+        return _id_name_dict(field)
```
A bit confusing here to use both `_id_name_dict()` and `id_name_dict()`, can we rename one?

elcct replied:
I changed it to `get_dict`. That wasn't best choice of a name.

reupen commented on datahub/search/dict_utils.py:
```diff
@@ -32,6 +32,26 @@ def _contact_dict(obj):
+        if len(fields) != 2:
+            raise ValueError("nested_field must be in 'nested_object.nested_field' format.")
+
+        related_object = getattr(obj, fields[0], None)
```
Why would the model instance not have the field? What are we protecting against by provided `None` as the third argument?

elcct replied:
This is true. It should have required attribute here.

reupen commented on datahub/search/dict_utils.py:
```diff
@@ -32,6 +32,26 @@ def _contact_dict(obj):
+        if related_object is None:
+            return None
+
+        field = getattr(related_object, fields[1], None)
```
Same question here about `getattr()` default value.

elcct replied:
It is when, as an example, `Contact` doesn't have `Company`, then related_object will be `None` and you won't be able to get `sector` out of it.

reupen replied:
The `if related_object is None` check just before this line will take care of that :)

elcct replied:
Of course :)

reupen commented on datahub/search/dict_utils.py:
```diff
@@ -32,6 +32,26 @@ def _contact_dict(obj):
     }
 
 
+def _nested_id_name_dict(nested_field):
```
It would be nice to differentiate this function from the other ones in this module, as this one returns a function while as the other ones return dicts.

elcct replied:
I have added `_computed` prefix to differentiate.

marcofucci commented:
Not really convinced by the sector being on the contact though. Why do we even need it? And if so maybe we could use a 2 level nested `contact.company.sector` instead?

elcct commented:
@marcofucci this is required for `Contact` collection. I added it like that to avoid having over complicated queries - if you wanted to search by sector, the resulted query would be quite big for what it would be doing, if we added another level of nesting.

marcofucci commented:
OK but in that case, you would have to add extra signals for managing when when a company and a sector change. In those cases the related contact has to be reindexed.
I would actually try to avoid these cases as much as possible as the signals and syncs to ES can add up quite quickly.
I have a similar `order.adviser.dit_team` scenario and I can oversea problems if we overuse this.

elcct commented:
We have to do something similar for Contact address if it is the same as company, With this PR I really only wanted to expose sector for FE and I would tackle signal stuff when working on address soon. Sadly I can't think of a way to avoid this this time.

marcofucci commented:
Sorry I meant that you need to catch post_save signal on company as somebody changes the company sector, you need to reindex the related contacts as you are denormalising the sector.
This should really be part of this PR as if you don't handle this, the indexed data might have inconsistent values.
